### PR TITLE
Improve failure handling during replication and promotion

### PR DESCRIPTION
**Pull request by roypaulin**
This is the first step to make online upgrade resilient.
We start with 2 critical steps:
- Replication: if replication fails, we are going to retry. We are going to delete the VerticaReplicator, clean up the annotation so that the operator can retry the next reconciler loop.
- Promotion: We are going to retry 3 times and if it still fails, we fail the online upgrade and log an error asking the user to revive_db/restore_backup.

### CODE REVIEW

**Comment by roypaulin** [comment]
I think the way we handle restore point during upgrade is too complicated and we should try to follow [VER-90388](https://jira.verticacorp.com/jira/browse/VER-90388)

File: pkg/controllers/vdb/onlineupgrade_reconciler.go [hunk 1]
```diff
@@ -223,9 +220,6 @@ func (r *OnlineUpgradeReconciler) Reconcile(ctx context.Context, _ *ctrl.Request
 		r.waitForReplicateToReplicaGroupB,
 		// replicate v_internal_tables.v_redirect_state to sandbox
 		r.copyRedirectStateToReplicaGroupB,
-		// Back up database after replication
-		r.postBackupDBAfterReplicationMsg,
```
Comment (cchen-vertica) [inline pkg/controllers/vdb/onlineupgrade_reconciler.go:227]: Remove these two functions below.

File: pkg/controllers/vdb/onlineupgrade_reconciler.go [hunk 2]
```diff
@@ -1113,11 +1117,27 @@ func (r *OnlineUpgradeReconciler) promoteSandboxToMainCluster(ctx context.Contex
...
 	actor := MakePromoteSandboxToMainReconciler(r.VRec, r.Log, r.VDB, sbPFacts, r.Dispatcher, r.VRec.Client)
 	r.Manager.traceActorReconcile(actor)
-	res, err := actor.Reconcile(ctx, &ctrl.Request{})
+	var res ctrl.Result
+	res, err = actor.Reconcile(ctx, &ctrl.Request{})
 	if verrors.IsReconcileAborted(res, err) {
-		return res, err
+		attempts := vmeta.GetOnlineUpgradePromotionAttempt(r.VDB.Annotations)
+		if attempts >= vmeta.OnlineUpgradePromotionMaxAttempts {
+			r.Log.Info("The last 3 sandbox promotion attempts have failed. Revive the database with the restorepoint", "sandboxName", r.sandboxName)
```
Comment (cchen-vertica) [inline pkg/controllers/vdb/onlineupgrade_reconciler.go]: Change "Revive the database with the restorepoint" to "Online upgrade cannot be done successfully. Please revive the database to its original state and try again."

File: pkg/controllers/vdb/onlineupgrade_reconciler.go [hunk 3]
```diff
@@ -1113,11 +1117,27 @@ func (r *OnlineUpgradeReconciler) promoteSandboxToMainCluster(ctx context.Contex
...
+			r.VRec.Eventf(r.VDB, corev1.EventTypeNormal, events.UpgradeFailed,
+				"Online upgrade has failed after 3 sandbox promotion attempts. Use restorepoint in archive %s to revive the database",
```
Comment (cchen-vertica) [inline pkg/controllers/vdb/onlineupgrade_reconciler.go]: Change "Use restorepoint in archive %s to revive the database" to "Please revive the database to its original state."

Comment (cchen-vertica) [inline pkg/controllers/vdb/onlineupgrade_reconciler.go]: Change 3 to %d

File: pkg/controllers/vdb/onlineupgrade_reconciler.go (hunk 2 above)
Comment (cchen-vertica) [inline pkg/controllers/vdb/onlineupgrade_reconciler.go]: Change "The last 3" to "All"

File: pkg/controllers/vdb/onlineupgrade_reconciler.go [hunk 4]
```diff
@@ -1113,11 +1117,27 @@ func (r *OnlineUpgradeReconciler) promoteSandboxToMainCluster(ctx context.Contex
...
+				vmeta.OnlineUpgradeArchiveAnnotation)
+			return ctrl.Result{}, errors.New("Sandbox promotion failed at least 3 times")
```
Comment (cchen-vertica) [inline pkg/controllers/vdb/onlineupgrade_reconciler.go]: Use %d to replace 3.

**Review by cchen-vertica — COMMENTED** [review]
Looks good. I only have minor comments.

File: pkg/controllers/vdb/onlineupgrade_reconciler.go [hunk 5]
```diff
@@ -1129,14 +1124,6 @@ func (r *OnlineUpgradeReconciler) promoteSandboxToMainCluster(ctx context.Contex
 	var res ctrl.Result
 	res, err = actor.Reconcile(ctx, &ctrl.Request{})
 	if verrors.IsReconcileAborted(res, err) {
```
Comment (cchen-vertica) [inline pkg/controllers/vdb/onlineupgrade_reconciler.go:1126]: We should only increase attempt number when there is an error. If a pod is not up and res is ctrl.Result{Requeue: true}, we don't want to increase the attempt number.

**Review by cchen-vertica — APPROVED** [review]

**Pull request merged**